### PR TITLE
python3Packages.notmuch2: add missing runtime dependency cffi

### PR DESCRIPTION
**Pull request by michaeladler**
## Description of changes

With this change it's possible to use notmuch2 after entering a nix-shell, e.g. `nix-shell -p python3Packages.notmuch2`. Previously, this would fail complaining about the missing cffi library.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/notmuch2/default.nix [hunk 1]
```diff
@@ -19,6 +19,8 @@ buildPythonPackage {
     python notmuch cffi
```
Comment (teto) [inline pkgs/development/python-modules/notmuch2/default.nix:19]: you can remove it from here than

**Comment by teto** [comment]
How can I reproduce failures solved by this ? is there a notmuch test we could use ?

File: pkgs/development/python-modules/notmuch2/default.nix (hunk 1 above)
Comment (michaeladler) [reply pkgs/development/python-modules/notmuch2/default.nix]: Done

**Comment by michaeladler** [comment]
It's very simple to reproduce this, just try to import a module:

```python
from notmuch2 import Database
```

Without my fix, this results in the following stacktrace:

```
Traceback (most recent call last):
...
  File "/nix/store/kzvd0m7f8qcxd7qw35a14br78inijm4l-python3.11-notmuch2-0.38/lib/python3.11/site-packages/notmuch2/__init__.py", line 45, in <module>
    from notmuch2 import _capi
ModuleNotFoundError: No module named '_cffi_backend'
```

I've tried to make the upstream tests work, but I'm stuck at this error:

```
E   ImportError: cannot import name '_capi' from partially initialized module 'notmuch2' (most likely due to a circular import) (/build/notmuch-0.38/bindings/python-cffi/notmuch2/__init__.py)
```

**Comment by teto** [comment]
thanks that indeed fixes it. Maybe you could set `python3 -c 'from notmuch2 import Database'` as a checkPhase (with `runHook preCheck`) ?

**Comment by michaeladler** [comment]
Yeah I had the same idea but it fails for the same reason as the tests:

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/build/notmuch-0.38/bindings/python-cffi/notmuch2/__init__.py", line 45, in <module>
    from notmuch2 import _capi
ImportError: cannot import name '_capi' from partially initialized module 'notmuch2' (most likely due to a circular import) (/build/notmuch-0.38/bindings/python-cffi/notmuch2/__init__.py)
```

I guess if we can figure this out then the tests might work as well and we can just run them.

**Review by teto — APPROVED** [review]
what if you run the tests in `installCheckPhase` instead ? let me know how you want to proceed I can merge this straightaway since it already improves the code and nixos release is closing in (and you can eventually ping me for a proper check if you wanna add one) or let this open until we have a proper check.

**Comment by michaeladler** [comment]
Same error, unfortunately. I'd say go ahead and merge this since I'll be rather busy in the next >= 2 weeks.

**Pull request merged**